pallets/order-book/src/lib.rs

```diff
@@ -0,0 +1,541 @@
+	pub struct Claim<T: Config> {
+		claiming_account: T::AccountId,
+		order_claiming: T::Hash,
+	}
+
+	#[pallet::storage]
+	pub type Orders<T: Config> = StorageMap<
+		_,
+		Twox64Concat,
+		T::Hash,
+		Order<T::Hash, T::AccountId, T::AssetCurrencyId, T::ForeignCurrencyBalance>,
+		ResultQuery<Error<T>::OrderNotFound>,
+	>;
+
+	/// Map of orders for a particular user
+	/// Used to query orders for a particular user using the
+	/// account id of the user as prefix
+	#[pallet::storage]
+	pub type UserOrders<T: Config> = StorageDoubleMap<
+		_,
+		Twox64Concat,
+		T::AccountId,
+		Twox64Concat,
+		T::Hash,
+		Order<T::Hash, T::AccountId, T::AssetCurrencyId, T::ForeignCurrencyBalance>,
+		ResultQuery<Error<T>::OrderNotFound>,
+	>;
+
+	/// Stores Nonce for orders placed
+	/// Given that Nonce is to ensure that all orders have a unique ID, we can
+	/// use just one Nonce, which means that we only have one val in storage,
+	/// and we don't have to insert new map values upon a new account/currency
+	/// order creation.
+	#[pallet::storage]
+	pub type NonceStore<T: Config> = StorageValue<_, T::Nonce, ValueQuery>;
+	#[pallet::storage]
+	pub type AssetPairOrders<T: Config> = StorageDoubleMap<
+		_,
+		Twox64Concat,
+		T::AssetCurrencyId,
+		Twox64Concat,
+		T::AssetCurrencyId,
+		BoundedVec<T::Hash, ConstU32<1_000_000>>,
+		ValueQuery,
+	>;
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		OrderCreated {
+			order_id: T::Hash,
+			creator_account: T::AccountId,
+			currency_in: T::AssetCurrencyId,
+			currency_out: T::AssetCurrencyId,
+			buy_amount: T::ForeignCurrencyBalance,
+			min_fullfillment_amount: T::ForeignCurrencyBalance,
+			sell_price_limit: T::ForeignCurrencyBalance,
+		},
+		OrderCancelled {
+			account: T::AccountId,
+			order_id: T::Hash,
+		},
+		OrderUpdated {
+			order_id: T::Hash,
+			account: T::AccountId,
+			buy_amount: T::ForeignCurrencyBalance,
+			sell_price_limit: T::ForeignCurrencyBalance,
+			min_fullfillment_amount: T::ForeignCurrencyBalance,
+		},
+		OrderFulfillment {
+			order_id: T::Hash,
+			placing_account: T::AccountId,
+			fulfilling_account: T::AccountId,
+			partial_fulfillment: bool,
+			fulfillment_amount: T::ForeignCurrencyBalance,
+			currency_in: T::AssetCurrencyId,
+			currency_out: T::AssetCurrencyId,
+			sell_price_limit: T::ForeignCurrencyBalance,
+		},
+	}
+
+	#[pallet::error]
+	#[derive(PartialEq)]
+	pub enum Error<T> {
+		AssetPairOrdersOverflow,
+		ConflictingAssetIds,
+		InsufficientAssetFunds,
+		InsufficientReserveFunds,
+		InvalidBuyAmount,
+		InvalidMinPrice,
+		InvalidAssetId,
+		OrderNotFound,
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T>
+	where
+		<T as frame_system::Config>::Hash: PartialEq<<T as frame_system::Config>::Hash>,
+	{
+		/// Create an order, with the minimum fulfillment amount set to the buy
+		/// amount, as the first iteration will not have partial fulfillment
```

Comment:
Oh I saw this after [this](https://github.com/centrifuge/centrifuge-chain/pull/1429/files#r1269004599) message. 

But what are your plans to realise not partial fulfillment, but the other way around: If a big order needs to eat many small orders?

> Assume there are many small ask orders at the best price (lowest ask). Now a big whale places a market order and buys all small ask orders. First, this often creates slippage, as you might know already.. Second, there might be not enough open ask orders to fill the big bid of the whale. Third, the whale needs to transact with many small fish accounts to fulfil the small orders. This leads to a high computational demand.

Reply:
Thats a good question.  The initial plan was to match orders of corresponding currency pairs and compatible prices and min fulfillment amounts, however there was some concern with the weight of matching those up, it's something I believe we plan on revisiting in the future.